core/transformer/src/PendingReferenceMap.ts

```diff
@@ -0,0 +1,119 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+/** @packageDocumentation
+ * @module Utils
+ */
+
+import { Id64String } from "@itwin/core-bentley";
+
+interface PendingReferenceProps {
+  referencer: Id64String;
+  referenced: Id64String;
+  isModelRef: boolean;
+}
+
+/**
+ * A reference relationships from an element, "referencer", to an element or its submodel, "referenced"
+ * @internal
+ */
+export class PendingReference implements PendingReferenceProps {
+  public referencer!: Id64String;
+  public referenced!: Id64String;
+  public isModelRef!: boolean;
+  constructor(props: PendingReferenceProps) {
+    Object.assign(this, props);
+  }
+  public toKey(): string {
+    return `${this.isModelRef ? "1" : "0"}${this.referencer}\x00${this.referenced}`;
+  }
+  private static keyPattern = /(?<isModelRef>0|1)(?<referencer>0x[0-9a-f])\x00(?<referenced>0x[0-9a-f])/;
+  public static fromKey(key: string): PendingReference {
+    const match = this.keyPattern.exec(key);
+    if (match === null || match.groups === undefined)
+      throw Error("invalid key passed");
+    return new PendingReference({
+      isModelRef: match.groups.isModelRef === "1",
+      referencer: match.groups.referencer,
+      referenced: match.groups.referenced,
+    });
+  }
+}
+
+/**
+ * a map that supports using PendingReferences objects as structural keys,
+ * as well as getting a list of referencers from a referencee (called referenced)
+ */
+export class PendingReferenceMap<T> implements Map<PendingReferenceProps, T> {
+  private _map = new Map<string, T>();
+  private _referencedToReferencers = new Map<Id64String, Set<Id64String>>();
+
+  public getReferencers(referenced: Id64String): Set<Id64String> {
+    let referencers = this._referencedToReferencers.get(referenced);
+    if (referencers === undefined) {
+      // lazy add an empty entry if there wasn't one
+      referencers = new Set();
+      this._referencedToReferencers.set(referenced, referencers);
+    }
+    return referencers;
+  }
+
+  /** if you have just the props of a pending reference, initialize a new class instance from it so we can use its convenience methods */
```

Comment:
It only has one non-static method, is this worth a separate allocation vs just calling a function to produce the key from the props?

Reply:
good point I'll switch to a function

Reply:
added a namespace for functions `PendingReference.toKey|fromKey`